File: go/prompt/number.go
package prompt

import (
	"fmt"
	"math"
	"regexp"
	"strconv"
	"strings"
)

// numberGrammar is the strict numeric-string grammar shared across all
// language implementations: optional sign, required integer part, optional
// fraction, optional exponent. It rejects "1_000", "3abc", "0x10", ".5",
// "5.", "", "+", and accepts "1e3", "  5  " (after trim), "3.5", "-2", "1E-3".
var numberGrammar = regexp.MustCompile(`^[+-]?\d+(\.\d+)?([eE][+-]?\d+)?$`)

// NumberConfig configures a numeric input prompt with optional bounds.
type NumberConfig struct {
	Message      string
	Default      *float64
	Min          *float64
	Max          *float64
	Step         *float64
	FloatAllowed bool
	KeepInput    *bool
	Validate     func(float64) error
	Filter       func(float64) float64
}

// Number prompts for a numeric value, validating against optional min/max bounds.
func Number(cfg NumberConfig) (float64, error) {
	if IsAgentMode() {
		// numberAgent handles validation, filter, and retry internally
		return numberAgent(cfg)
	}
	result, err := numberTerminal(cfg)
	if err != nil {
		return 0, err
	}
	if cfg.Validate != nil {
		if err := cfg.Validate(result); err != nil {
			return 0, fmt.Errorf("%w: %v", ErrValidation, err)
		}
	}
	if cfg.Filter != nil {
		result = cfg.Filter(result)
	}
	return result, nil
}

func numberAgent(cfg NumberConfig) (float64, error) {
	payload := map[string]any{
		"type":          "number",
		"message":       cfg.Message,
		"default":       cfg.Default,
		"min":           cfg.Min,
		"max":           cfg.Max,
		"num_step":      cfg.Step,
		"float_allowed": cfg.FloatAllowed,
	}
	raw, err := AgentPromptWithRetry(payload, func(answer any) (any, error) {
		result, err := validateNumber(answer, cfg)
		if err != nil {
			return nil, err
		}
		if cfg.Validate != nil {
			if err := cfg.Validate(result); err != nil {
				return nil, fmt.Errorf("%w: %v", ErrValidation, err)
			}
		}
		if cfg.Filter != nil {
			result = cfg.Filter(result)
		}
		return result, nil
	})
	if err != nil {
		return 0, err
	}
	val, ok := raw.(float64)
	if !ok {
		return 0, fmt.Errorf("%w: expected float64, got %T", ErrValidation, raw)
	}
	return val, nil
}

func numberKeepInput(cfg NumberConfig) bool {
	if cfg.KeepInput != nil {
		return *cfg.KeepInput
	}
	return true // default true
}

func numberTerminal(cfg NumberConfig) (float64, error) {
	t := DefaultTheme
	scanner := getTerminalScanner()
	keepInput := numberKeepInput(cfg)
	for {
		suffix := ""
		if cfg.Default != nil {
			suffix = fmt.Sprintf(" (%g)", *cfg.Default)
		}
		fmt.Printf("%s %s%s: ", t.SymQuestion, cfg.Message, suffix)
		if !scanner.Scan() {
			return 0, ErrAborted
		}
		raw := scanner.Text()
		if raw == "" && cfg.Default != nil {
			fmt.Printf("%s %s %g\n", t.SymSuccess, cfg.Message, *cfg.Default)
			return *cfg.Default, nil
		}
		result, err := validateNumber(raw, cfg)
		if err != nil {
			fmt.Printf("  %s\n", err)
			if keepInput {
				parsed, parseErr := strconv.ParseFloat(raw, 64)
				if parseErr == nil {
					cfg.Default = &parsed
				}
			}
			continue
		}
		fmt.Printf("%s %s %g\n", t.SymSuccess, cfg.Message, result)
		return result, nil
	}
}

// isASCIISpace reports whether r is an ASCII whitespace character (space, tab,
// newline, carriage return, vertical tab, form feed).
func isASCIISpace(r rune) bool {
	switch r {
	case ' ', '\t', '\n', '\r', '\v', '\f':
		return true
	default:
		return false
	}
}

func validateNumber(v any, cfg NumberConfig) (float64, error) {
	var num float64
	switch val := v.(type) {
	case float64:
		if math.IsNaN(val) || math.IsInf(val, 0) {
			return 0, fmt.Errorf("%w: Not a valid number", ErrValidation)
		}
		num = val
	case int:
		num = float64(val)
	case string:
		// Trim leading/trailing ASCII whitespace, then the remainder MUST
		// fully match the numeric grammar before parsing.
		trimmed := strings.TrimFunc(val, isASCIISpace)
		if !numberGrammar.MatchString(trimmed) {
			return 0, fmt.Errorf("%w: Not a valid number: %q", ErrValidation, val)
		}
		parsed, err := strconv.ParseFloat(trimmed, 64)
		if err != nil || math.IsNaN(parsed) || math.IsInf(parsed, 0) {
			return 0, fmt.Errorf("%w: Not a valid number: %q", ErrValidation, val)
		}
		num = parsed
	case nil:
		if cfg.Default != nil {
			return *cfg.Default, nil
		}
		return 0, fmt.Errorf("%w: Expected a number, got nil", ErrValidation)
	default:
		return 0, fmt.Errorf("%w: Expected a number, got %T", ErrValidation, v)
	}

	if !cfg.FloatAllowed && num != math.Trunc(num) {
		return 0, fmt.Errorf("%w: Decimal numbers are not allowed", ErrValidation)
	}
	if cfg.Min != nil && num < *cfg.Min {
		return 0, fmt.Errorf("%w: must be at least %g", ErrValidation, *cfg.Min)
	}
	if cfg.Max != nil && num > *cfg.Max {
		return 0, fmt.Errorf("%w: must be at most %g", ErrValidation, *cfg.Max)
	}
	if cfg.Step != nil {
		base := 0.0
		if cfg.Min != nil {
			base = *cfg.Min
		}
		remainder := math.Mod(num-base, *cfg.Step)
		if math.Abs(remainder) > 1e-9 && math.Abs(remainder-*cfg.Step) > 1e-9 {
			return 0, fmt.Errorf("%w: must be a multiple of %v (from %v)", ErrValidation, *cfg.Step, base)
		}
	}
	return num, nil
}

File: python/src/inquirer_ai/prompts/number.py
from __future__ import annotations

import math
import re
from typing import Any

from prompt_toolkit import prompt as pt_prompt
from prompt_toolkit.formatted_text import FormattedText

from inquirer_ai.exceptions import ValidationError
from inquirer_ai.prompts.base import BasePrompt
from inquirer_ai.theme import RESET, get_theme

# Numeric-string grammar (R2): optional sign; required integer part; optional
# .fraction; optional exponent. Rejects "1_000", "3abc", "0x10", ".5", "5.",
# "", "+". Accepts "1e3", "3.5", "-2", "1E-3".
_NUMBER_RE = re.compile(r"^[+-]?\d+(\.\d+)?([eE][+-]?\d+)?$")
# ASCII whitespace to strip (leading/trailing) before matching.
_ASCII_WS = " \t\n\r\f\v"


class NumberPrompt(BasePrompt[int | float]):
    def __init__(
        self,
        message: str,
        *,
        min: int | float | None = None,
        max: int | float | None = None,
        step: int | float | None = None,
        float_allowed: bool = True,
        keep_input: bool = True,
        **kwargs: Any,
    ) -> None:
        super().__init__(message, **kwargs)
        self.min = min
        self.max = max
        self.step = step
        self.float_allowed = float_allowed
        self.keep_input = keep_input

    @property
    def prompt_type(self) -> str:
        return "number"

    def _validate_answer(self, value: Any) -> int | float:
        # 1) null + default present -> default.
        if value is None and self.default is not None:
            return self.default
        num: int | float
        if isinstance(value, bool):
            # 4) bool is not a number.
            raise ValidationError(f"Expected a number, got {type(value).__name__}")
        elif isinstance(value, (int, float)):
            # 2) JSON number (not boolean) -> use it.
            num = value
        elif isinstance(value, str):
            # 3) trim leading/trailing ASCII whitespace then enforce the grammar.
            trimmed = value.strip(_ASCII_WS)
            if not _NUMBER_RE.match(trimmed):
                raise ValidationError(f"Not a valid number: {value!r}")
            # Parse with the native float parser; return an int (language-idiomatic)
            # for pure-integer forms (no fraction/exponent). Numeric value is
            # identical across languages either way.
            num = float(trimmed) if any(ch in trimmed for ch in ".eE") else int(trimmed)
        else:
            # 4) other type.
            raise ValidationError(f"Expected a number, got {type(value).__name__}")
        # 5) reject non-finite (NaN/Inf).
        if isinstance(num, float) and not math.isfinite(num):
            raise ValidationError(f"Not a valid number: {value!r}")
        # 6) if !float_allowed: require integral else error, then coerce to int.
        if not self.float_allowed:
            if isinstance(num, float) and not num.is_integer():
                raise ValidationError("Decimal numbers are not allowed")
            num = int(num)
        if self.min is not None and num < self.min:
            raise ValidationError(f"Must be at least {self.min}")
        if self.max is not None and num > self.max:
            raise ValidationError(f"Must be at most {self.max}")
        if self.step is not None:
            base = self.min if self.min is not None else 0
            remainder = (num - base) % self.step
            if abs(remainder) > 1e-9 and abs(remainder - self.step) > 1e-9:
                raise ValidationError(f"Must be a multiple of {self.step} (from {base})")
        return num

    def _to_agent_dict(self) -> dict[str, Any]:
        d = super()._to_agent_dict()
        d["min"] = self.min
        d["max"] = self.max
        d["num_step"] = self.step
        d["float_allowed"] = self.float_allowed
        return d

    def _execute_terminal(self) -> int | float:
        t = get_theme()
        suffix = f" ({self.default})" if self.default is not None else ""
        retry_default: str | None = None
        while True:
            message = FormattedText(
                [
                    (t.pt(t.question), f"{t.sym_question} "),
                    ("bold", f"{self.message}{suffix}: "),
                ]
            )
            raw = pt_prompt(message, default=retry_default or "")
            if not raw and self.default is not None:
                return self.default
            try:
                result = self._validate_answer(raw)
            except ValidationError as e:
                print(f"{t.ansi(t.error)}  {e}{RESET}")
                retry_default = raw if self.keep_input else None
                continue
            # Run user-provided validation
            error = self._run_user_validation(result)
            if error:
                print(f"{t.ansi(t.error)}  {error}{RESET}")
                retry_default = raw if self.keep_input else None
                continue
            return result
